docroot/modules/custom/foia_users/foia_users.module

```diff
@@ -40,63 +40,83 @@ function _office_user_nid_access($node, $account) {
 function foia_users_node_access($node, $op, $account) {
 
   // Restrict editing node based on matching of 'field_agency' entity reference.
-  if ($node->bundle() == 'foia_office') {
-
+  if ($node->bundle() == 'agency_component') {
+    // Get the current user's roles.
     $user = User::load(\Drupal::currentUser()->id());
     $user_roles = $user->getRoles();
 
-    if (in_array("foia_office_creator", $user_roles) || in_array("foia_office_reviewer", $user_roles)) {
+    if (in_array("agency_manager", $user_roles)) {
+      switch ($op) {
+        case 'update':
+          // Check if the user has the proper Agency association.
+          return _agency_user_nid_access($node, $account);
 
-      if ($op == "delete") {
-        // Allow creation before entity ref is set.
-        if ($node->get('field_agency')->isEmpty()) {
-          return AccessResult::allowed()
+        case 'delete':
+          return AccessResult::forbidden()
             ->cachePerUser()
             ->addCacheableDependency($node);
-        }
-        else {
-          return _office_user_nid_access($node, $account);
-        }
-      }
-      elseif ($op == "update") {
-        return _office_user_nid_access($node, $account);
+
+        default:
+          return AccessResult::neutral();
       }
-      else {
+    }
+  }
+  return AccessResult::neutral();
+}
+
+/**
+ * Implements hook_ENTITY_TYPE_access().
+ */
+function foia_users_taxonomy_term_access(EntityInterface $entity, $operation, AccountInterface $account) {
+  // Get the term ID.
+  $term_id = $entity->id();
+  // Get the current user's roles.
+  $user = User::load(\Drupal::currentUser()->id());
+  $user_roles = $user->getRoles();
+  // Get the current user's agency.
+  $user_agency = $user->get('field_agency')->target_id;
+
+  if ($term_id === $user_agency && in_array('agency_manager', $user_roles)) {
+    switch ($operation) {
+      case 'update':
         return AccessResult::allowed()
           ->cachePerUser()
-          ->addCacheableDependency($node);
-      }
+          ->addCacheableDependency($entity);
 
-    }
+      case 'delete':
+        return AccessResult::forbidden()
+          ->cachePerUser()
+          ->addCacheableDependency($entity);
 
+      default:
+        return AccessResult::neutral();
+    }
   }
   return AccessResult::neutral();
 }
 
 /**
- * Custom submit handler for office value.
+ * Custom submit handler for agency value.
  */
-function _form_validation_foiaoffice($form, $form_state) {
-
-  if ($form_state->hasValue('field_foia_office')) {
-
-    $role_array = ["foia_office_creator", "foia_office_reviewer"];
+function _form_validation_foia_agency($form, $form_state) {
+  if ($form_state->hasValue('field_agency')) {
```

Comment:
If a condition to showing the error message below is that the form has a value for `field_agency`, it will never display, will it?

Reply:
`field_agency` is the entity reference field to the Agency taxonomy term. The `hasValue` evaluates to true if that field exists. It works more like `array_key_exists()`.

Reply:
You're right! this worked.